style.ts

```diff
@@ -20,41 +26,35 @@ export interface AbstractExpression {
   type: string;
 };
 
-/**
- * Expression that evaluates to the given value.
- */
-export interface LiteralValue<T> extends AbstractExpression {
-  type: 'literal';
-  value: T;
-};
 
 /**
  * Expression that evaluates to the value of the given property.
  */
 export interface PropertyName extends AbstractExpression {
   type: 'property';
-  name: string;
+  name: Expression<string>;
 };
 
 /**
  * Expression that evaluates to the result of a function
  * call on a list of argument expressions.
  */
-export interface FunctionCall extends AbstractExpression {
-  type: 'functioncall';
-  name: string;
-  args: Expression[];
+export interface FunctionCall<T> extends AbstractExpression {
+  type: T extends string ? 'stringfunction' :
+    T extends number ? 'numberfunction' :
+    T extends boolean ? 'booleanfunction' :
+    'voidfunction';
+  name: T extends string ? GeoStylerStringFunction['name'] :
+    T extends number ? GeoStylerNumberFunction['name'] :
+    T extends boolean ? GeoStylerBooleanFunction['name'] :
+    unknown ;
```

Comment:
Would it be nicer if there was a type
```
type GeostylerFunction = GeoStylerStringFunction | GeoStylerNumberFunction | GeostylerBooleanFunction;
```
?
Then this could be replaced by:|
```
name: GeoStylerFunction['name'];
```